rust/geoarrow-array/src/cast.rs

```diff
@@ -820,6 +820,950 @@ macro_rules! downcast_geoarrow_array {
     };
 }
 
+/// Two argument version of the `downcast_geoarrow_array!` macro.
+/// Downcast any combination of two [GeoArrowArray] to a concrete-typed array based on its [`GeoArrowType`].
+#[macro_export]
+macro_rules! downcast_geoarrow_array_two_args {
+    ($array1:ident, $array2:ident, $fn:expr) => {
+        match $array1.data_type() {
```

Comment:
I think it would be a lot cleaner here and a lot easier to read if we had
```rs
match ($array1.data_type(), $array2.data_type())
```
then we don't need two levels of nesting.